### ISSUE TITLE
Test Loss Calculation across TPU Nodes

### ISSUE THREAD

**Issue by KaleabTessera** [pytorch/xla#2550]
## ❓ Questions and Help
My test loss values seem to be inaccurate when aggregating across tpu nodes. Am I doing something wrong?
 
I am calculating test loss across TPU nodes using a similiar approach to an example in the torch xla code https://github.com/pytorch/xla/blob/56138cf7b29dc20ed9b0ca5934b91d1cf9a72b70/test/test_train_mp_mnist.py#L148

Here is the my test code:
```
def run_test_tpu(model, device, test_loader, epoch, exp=None, conv_type=None):
    tracker = xm.RateTracker()
    model.eval()
    test_loss = 0
    top1 = AverageMeter('Acc@5', ':6.2f')
    top5 = AverageMeter('Acc@5', ':6.2f')

    with torch.no_grad():
        for data, target in test_loader:
            output = model(data)
            # sum up batch loss
            test_loss += F.cross_entropy(output,
                                         target, reduction='sum').item()

            # Accuracy
            correct_top1, correct_top5 = accuracy(output, target, topk=(1, 5))
            top1.update(correct_top1[0], data.size(0))
            top5.update(correct_top5[0], data.size(0))

    test_loss /= test_loader.__len__()
    test_accuracy = top1.avg.item()
    test_accuracy_5 = top5.avg.item()

    test_accuracy_agg = xm.mesh_reduce(
        'test_accuracy_agg', test_accuracy, np.mean)
    test_accuracy_agg_5 = xm.mesh_reduce(
        'test_accuracy_agg_5', test_accuracy_5, np.mean)
    test_loss_agg = xm.mesh_reduce(
        'test_loss_agg', test_loss, np.mean)

    return test_accuracy_agg, test_accuracy_agg_5, test_loss_agg
```

So test accuracy seems to work and be correct, but test loss is a really high, unaccurace number. 

The train version of this seems to work:
```
def run_train_tpu(model, device, train_loader, optimizer, epoch, exp, printTrain, conv_type, l1, mask_grad, gradclip):
    tracker = xm.RateTracker()
    model.train()
    running_loss = 0.0
    top1 = AverageMeter('Acc@5', ':6.2f')
    top5 = AverageMeter('Acc@5', ':6.2f')
    for batch_idx, (data, target) in enumerate(train_loader):
        optimizer.zero_grad()
        output = model(data)
        loss = F.cross_entropy(output, target)
        if(l1):
            reg = l1
            l1_norm = torch.zeros(1).to(device)
            for name, param in model.named_parameters():
                if 'weight' in name:
                    l1_norm = l1_norm + (reg * torch.sum(torch.abs(param)))
            loss = loss + l1_norm

        loss.backward()

        if(gradclip and gradclip > 0):
            nn.utils.clip_grad_norm_(model.parameters(), gradclip)

        xm.optimizer_step(optimizer)
        running_loss += loss.item()

        # Accuracy
        correct_top1, correct_top5 = accuracy(output, target, topk=(1, 5))
        top1.update(correct_top1[0], data.size(0))
        top5.update(correct_top5[0], data.size(0))

    # Normalizing the loss by the total number of train batches
    train_loss = running_loss / train_loader.__len__()
    train_accuracy = top1.avg.item()
    train_accuracy_5 = top5.avg.item()

    train_accuracy_agg = xm.mesh_reduce(
        'train_accuracy_agg', train_accuracy, np.mean)
    train_accuracy_agg_5 = xm.mesh_reduce(
        'train_accuracy_agg_5', train_accuracy_5, np.mean)
    train_loss_agg = xm.mesh_reduce(
        'train_loss_agg', train_loss, np.mean)

    return train_accuracy_agg, train_loss_agg
```

Here are my loaders: 
```
train_dataset, test_dataset = SERIAL_EXEC.run(lambda: get_dataset(dataset=dataset,data_folder=data_folder, transform_train=transform_train, transform_test=transform_test))

train_sampler = torch.utils.data.distributed.DistributedSampler(
    train_dataset,
    num_replicas=xm.xrt_world_size(),
    rank=xm.get_ordinal(),
    shuffle=True)

test_sampler = torch.utils.data.distributed.DistributedSampler(
    test_dataset,
    num_replicas=xm.xrt_world_size(),
    rank=xm.get_ordinal(),
    shuffle=False)

train_loader = torch.utils.data.DataLoader(
    train_dataset,
    batch_size=batch_size,
    sampler=train_sampler,
    num_workers=4,
    drop_last=True)

test_loader = torch.utils.data.DataLoader(
    test_dataset,
    batch_size=test_batch_size,
    sampler=test_sampler,
    shuffle=False,
    num_workers=4,
    drop_last=True)

```

**Comment by taylanbil** [comment]
First, I would advise against using `.item()` in the loop, as it would trigger an early evaluation. consequently, test_loss will be a tensor inside the loop, and you could call `.item()` on it at the end as `test_loss = test_loss.item() / test_loader.__len__()`.

As to why the loss is high, I see that the reduction in `F.cross_entropy` is sum. This requires normalization by the number of total samples processed. So in the current state, the code is adding up all the loss values per **sample** and dividing by the number of **batches**. The correct computation of loss would require one of the two:

- normalize by the total number of samples, not `test_loader.__len__`. In order to do so, you could keep a running count of batch sizes in the loop.
- use reduction `mean` instead of `sum`.

**Comment by taylanbil** [comment]
@KaleabTessera did the above comment help your issue?

**Comment by KaleabTessera** [comment]
I ended up doing option 1 (`normalize by the total number of samples, not test_loader.__len__. In order to do so, you could keep a running count of batch sizes in the loop.`) as follows:
<pre><code>
def run_test_tpu(model, device, test_loader, epoch, exp=None, conv_type=None):
    tracker = xm.RateTracker()
    model.eval()
    test_loss = 0
   <b> total_batch_size = 0 </b> 
    top1 = AverageMeter('Acc@5', ':6.2f')
    top5 = AverageMeter('Acc@5', ':6.2f')

    with torch.no_grad():
        for data, target in test_loader:
            output = model(data)
            # sum up batch loss
            test_loss += F.cross_entropy(output,
                                         target, reduction='sum')

            # Accuracy
            correct_top1, correct_top5 = accuracy(output, target, topk=(1, 5))
            top1.update(correct_top1[0], data.size(0))
            top5.update(correct_top5[0], data.size(0))

           <b> total_batch_size += data.size(0) </b>

   <b> test_loss = test_loss.item() / total_batch_size </b>
    test_accuracy = top1.avg.item()
    test_accuracy_5 = top5.avg.item()

    test_accuracy_agg = xm.mesh_reduce(
        'test_accuracy_agg', test_accuracy, np.mean)
    test_accuracy_agg_5 = xm.mesh_reduce(
        'test_accuracy_agg_5', test_accuracy_5, np.mean)
    test_loss_agg = xm.mesh_reduce(
        'test_loss_agg', test_loss, np.mean)

    if exp is not None:
        exp.log_metric("loss", test_loss_agg, epoch=epoch)
        exp.log_metric("accuracy", test_accuracy_agg, epoch=epoch)
        exp.log_metric("accuracy_top5", test_accuracy_agg_5, epoch=epoch)

    return test_accuracy_agg, test_accuracy_agg_5, test_loss_agg
</code></pre>

Thanks! @taylanbil . Can a create a PR to put calculating test_loss & train_loss in some of the code examples?

**Comment by taylanbil** [comment]
[The imagenet example](https://github.com/pytorch/xla/blob/master/test/test_train_mp_imagenet.py) does compute the test and train loss. The root cause of the issue here wasn't TPU specific, it was just the wrong arithmetic being done, yielding an unusually high number. As to not calling `.item()` inside the loop, that's [mentioned in the docs](https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#known-performance-caveats). Feel free to send a PR though if you feel like the existing examples can be improved.

Closing this since the issue is resolved, feel free to re-open if needed.
